Makefile

```diff
@@ -0,0 +1,128 @@
+# Setup Tasks
+# --------------------------------------------------------------------------------------------------
+
+install: ## install dependencies and redact github token
+	cd project && go get -d -v ./... 2>&1 | sed -e "s/[[:alnum:]]*:x-oauth-basic/redacted/"
+
+test: ## run tests on package and all subpackages
+	cd project && go test $(LDFLAGS) -v -race -tags integration ./...
+
+lint: ## run the linter
+	cd project && golangci-lint run --deadline=2m
+
+# --------------------------------------------------------------------------------------------------
+# Build Tasks
+# --------------------------------------------------------------------------------------------------
+
+build-app:
+ifneq ("$(wildcard project/main.go)","")
+	cd project && CGO_ENABLED=0 go build $(LDFLAGS) -o ../bin/$(MAIN_IMAGE_NAME) -a .
+endif
+
+cmd_sources = $(dir $(wildcard ./project/cmd/*/main.go))
+cmds = $(foreach source,$(cmd_sources),$(patsubst %/,%,$(subst ./project/cmd,./bin,$(source))))
+
+define go-build
+	cd ./$< && CGO_ENABLED=0 go build $(LDFLAGS) -o ./../../../$@ -a .
+endef
+
+./bin/%: ./project/cmd/% ## build individual command
+	$(go-build)
+
+build-commands: $(cmds) ## build all commands
+
+build-all: build-app build-commands
+
+# --------------------------------------------------------------------------------------------------
+# Docker Build Tasks
+# --------------------------------------------------------------------------------------------------
+
+docker_commands = $(foreach source,$(cmd_sources),$(subst /,,$(subst ./project/cmd/,docker-build-cmd-,$(source))))
+
+define docker-build
+	docker build -f Dockerfile.project -t $(DOCKER_BASE_NAME)/$(image_name):$(CIRCLE_SHA1) . --build-arg EXECUTABLE=$(<F)
+	docker save --output ./images/$(image_name).tar $(DOCKER_BASE_NAME)/$(image_name):$(CIRCLE_SHA1)
+endef
+
+docker-build-app: ## build docker image for main app
+ifneq ("$(wildcard project/main.go)","")
+	docker build -f Dockerfile.project -t $(DOCKER_BASE_NAME)/$(MAIN_IMAGE_NAME):$(CIRCLE_SHA1) . --build-arg EXECUTABLE=$(MAIN_IMAGE_NAME)
+	docker save --output ./images/$(MAIN_IMAGE_NAME).tar $(DOCKER_BASE_NAME)/$(MAIN_IMAGE_NAME):$(CIRCLE_SHA1)
+endif
+
+docker-build-cmd-%: image_name = $(MAIN_IMAGE_NAME)-$(subst docker-build-cmd-,,$@)
+docker-build-cmd-%: ./bin/% ## build docker image for one command
+	$(docker-build)
+
+docker-build-commands: $(docker_commands)
+
+ensure-static: ## ensures that a static folder exists in the project
+	mkdir -p project/static
+
+docker-build-all: ensure-static build-all docker-build-app docker-build-commands
+
+# --------------------------------------------------------------------------------------------------
+# Docker Push Tasks
+# --------------------------------------------------------------------------------------------------
+
+image_sources = $(sort $(shell find ./images -mindepth 1 -maxdepth 1 -exec basename {} \;))
+images = $(foreach image,$(image_sources),docker-push-image-$(subst .tar,,$(image)))
```

Comment:
Why do we build images to a tar?

Reply:
this is to accommodate the current setup where we use two different contexts one with github credentials and one with docker ones. We persist the images to the workspace at the end of a build job and then load them in the push job. If we can add the github token to our context with docker credentials, then we can simplify this.